README.md
# Pohjola III
Pohjola III is a tactical turn-based fighting game with RPG mechanics.  
Develop your character with equipment, perks and skills in order to face fantasy creatures.

## Features
- Turn-based combat, with a speed mechanic.
  * Every character has a speed stat, which increases the rate at which their action bar fills.
  * The action bar is used to perform actions, such as attacking, using items, etc and is depleted fully on use.
- Hotbar with 6 slots for skills.
  * Choose your skills carefully before battle.
- Perks and skills.
  * Perks are passive abilities that can be unlocked by leveling up.
  * Skills are active abilities that can be used from the hotbar.
- Equipment.
  * Equip weapons, armor and accessories to increase your stats.
- Challenges.
  * At game start, you can choose from many different challenges.
  * Challenges can greatly affect how you approach the game.

## Development
Current game version: 0.2.1
[CHANGELOG.md](https://github.com/WelehoBRUDER/Pohjola-III/blob/main/CHANGELOG.md)  
The game's development may be sporadic, with long periods of inactivity and short bursts of activity.  
Any suggestions or found bugs can be reported in the [issues](https://github.com/WelehoBRUDER/Pohjola-III/issues) tab.


CHANGELOG.md
# Version 0.2.1 ~ 2023.4.9

## New features & additions
- Added new stat ``accuracy``.
  - Accuracy is a stat that affects your chance to hit.
  - It works by decreasing the enemy's dodge chance.
- Added new skill ``Vile Strikes`` for all classes.
  - Vile Strikes is a skill that increases your accuracy and crit rate.
  - Min: 40% acc, 40% crit
  - Max: 100% acc, 100% crit
- Added easier difficulty settings.
  - Enemy damage can be set to 80%/60%/40%.
  - Enemy health can be set to 80%/60%/50%.
  - Healing effectiviness can be set to 125%/150%/175%.
  - Mana regeneration can be set to 120%/150%/200%.
  - These settings will lower score multiplier.
- Added new items:
  - "Blade of Hero" (epic sword)
  - "Signet of Precision" (rare ring)
  - "Signet of Power" (rare ring)
  - "Signet of Health" (rare ring)
- Added button to go back to main menu
  - Next to load from file button in saves screen.

## Changes
- "Battle Aura"
  - Increases accuracy by 15/20/25/30
  - Duration increased from 7.5 --> 10 seconds.
  - Cooldown increased from 12 --> 16 seconds.
  - Upgraded durations are now 11/12/13 seconds.
- Some debuff effects now also lower accuracy.
- Made escape more universal.
- Added exit buttons to menu screens.
- Changed some internal parameter names.

# Version 0.2.0 ~ 2023.4.1

## New features & additions
- Added new enemies:
  - Minotaur Sage
  - Minotaur Captain
  - Minotaur King (Oros, King of Minor)
- Added new dungeon: `Fortress of Minor`
  - Unlocked after defeating the orc chieftain.
  - Fight boss "Minotaur King" to complete.
- Added new items:
  - Minotaur armor set.
  - Ring of Bull Power.
  - New potion tiers.
  - Blessed Graves.
  - Five talismans, each increase a stat by 3.
- Added search & filter to all inventory screens (including store).
- Added command "lose" to the game.
- Playtime is finally tracked. (It no longer stays at 0)
- The game now tracks score, and challenges improve the multiplier.

## Changes
- Added highlight to perk tree when hovering over a perk.
  - Highlights perks that are required to unlock the hovered perk.
  - Will not highlight already owned perks.
- Dodge text is now silver instead of black.
- Added more clarity between dungeons and floors / stages.
- Slightly buffed Orc Chief Armor.
- Changed save slot font and some other minor UI changes.
- Reworked the statistics screen.
  - Added more statistics.
  - Statistics are divided into 7 categories.
- Scrollbar is now visible in more screens.
- Slightly improved ui scaling.

## Bug fixes
- Fixed enemies attacking after death.
- Fixed enemies attacking multiple times in a row.
- Fixed lock-on targeting not working properly.
- Fixed overwriting saves not working.
- Fixed save files acting weird when loading from a hardcore save.
- Fixed "most healing done" not being tracked.

# Version 0.1.4 ~ 2023.3.19

## New features & additions
- Added level 5 & 10 perks for each class. (24 new perks in total!)
- Added new skill "Poison cloud" for the rogue class.
- Added "ignore" command to the game.

## Bug fixes
- Fixed certain old saves breaking upon load.
- Fixed not getting hp & mp recovery during level up with "no recovery" challenge active.

# Version 0.1.3 ~ 2023.3.16

## New features & additions
- Added classes.
  - Currently there are 4 classes: warrior, rogue, mage and paladin.
- Added purchasable perks for each class.
  - Perks are unlocked at certain levels and cost money.
  - Perks can also require certain stats to be unlocked.

## Changes
- See breakdown of your modifiers in character screen (tooltip when hovering).
- Stun resistance now reduces the duration of stuns instead of resisting them.

# Version 0.1.2 ~ 2023.3.14

## New features & additions
- Added new dungeon: `Abandoned Fort`
  - Unlocked after defeating the troll.
  - Fight boss "Orc Chieftain" to complete.
- Added orc related items.
  - Orc chief armor set.
  - Orc berserker axes.
- Added "killall" command to the game.
- Added "enter" command to the game.
- Added "leave" command to the game.

## Bug fixes
- Fixed text overflowing from notifications.
- Fixed stun resistance doing nothing.

# Version 0.1.1 ~ 2023.3.12

## New features & additions
- Added target lock on mechanic to settings.
- Added "fight" command to the game.
- Added the "ranger" armor set. 
  - This armor set is a bit weaker than the "plate" armor set, but it has a chance to dodge attacks.
- Added some missing localization.
## Bug fixes
- Fixed being able to upgrade the first perk, despite it having no levels. 
  - This happened because the max level check was done after a requirement check  
  which the first perk passes by default.
- Fixed not being able to unequip armor.

# Version 0.1 ~ 2023.3.11
This is the first proper version of Pohjola III.  
The game is still in early development, so expect bugs and missing features.

## New features & additions
- Dungeons mechanics are mostly working.
  - The game currently has 1 dungeon.

### This version does not have any changes as there are no previous versions.


script/enemies.js
const Enemies = {
  "skeleton_warrior": {
    id: "skeleton_warrior",
    name: "Skeleton Warrior",
    sprites: "gfx/enemies/skeleton_warrior",
    stats: {
      str: 1,
      vit: 3,
      agi: 1,
      int: 1,
      wis: 1,
      ap: 0,
      hp: 0,
      mp: 0,
      healL: 40,
      critChance: 5,
      critDmg: 50
    },
    resistances: {
      burning: 0,
      bleeding: 0,
      stun: 0,
      power: 0,
      defense: 0,
      freezing: 0
    },
    skills: {
      armorer: 5,
      heavy_weapons: 0,
      light_weapons: 5,
      shield: 5,
      dodge: 0,
      dexterity: 0
    },
    inventory: [
      new Item({...items.weak_healing_gem})
    ],
    equipment: {
      weapon: new Item(items.rusty_short_sword),
      shield: new Item(items.wooden_shield),
      head: {},
      body: {},
      legs: {},
      ring: {}
    },
    abilities: [
      new Ability(Abilities.regular_attack),
      new Ability(Abilities.shield_bash),
    ],
    level: {
      lvl: 1,
      pointsPerLvl: 3,
      xpGive: 10
    },
    statuses: [
    ],
    stat_template: "defender",
    ai_prefers: {
      regular: 25,
      debuff: 0
    },
    lootTable: "skeleton1"
  },
  "skeleton_archer": {
    id: "skeleton_archer",
    name: "Skeleton Archer",
    sprites: "gfx/enemies/skeleton_archer",
    stats: {
      str: 3,
      vit: 1,
      agi: 3,
      int: 1,
      wis: 1,
      ap: 0,
      hp: 0,
      mp: 0,
      healL: 30,
      critChance: 10,
      critDmg: 75
    },
    resistances: {
      burning: 0,
      bleeding: 0,
      stun: 0,
      power: 0,
      defense: 0,
      freezing: 0
    },
    skills: {
      armorer: 0,
      heavy_weapons: 10,
      light_weapons: 0,
      shield: 0,
      dodge: 0,
      dexterity: 0
    },
    inventory: [
      new Item({...items.weak_healing_gem})
    ],
    equipment: {
      weapon: new Item(items.bent_wooden_bow),
      shield: {},
      head: {},
      body: {},
      legs: {},
      ring: {}
    },
    level: {
      lvl: 1,
      pointsPerLvl: 3,
      xpGive: 15
    },
    abilities: [
      new Ability(Abilities.regular_attack),
      new Ability(Abilities.sharp_stroke)
    ],
    statuses: [],
    stat_template: "ranger",
    ai_prefers: {
      regular: 0,
      debuff: 50
    },
    lootTable: "skeleton2"
  },
  "skeleton_knight": {
    id: "skeleton_knight",
    name: "Skeleton Knight",
    sprites: "gfx/enemies/skeleton_knight",
    stats: {
      str: 8,
      vit: 6,
      agi: 6,
      int: 1,
      wis: 1,
      ap: 0,
      hp: 0,
      mp: 0,
      healL: 35,
      critChance: 5,
      critDmg: 50
    },
    resistances: {
      burning: 0,
      bleeding: 0,
      stun: 0,
      power: 0,
      defense: 0,
      freezing: 0
    },
    skills: {
      armorer: 10,
      heavy_weapons: 10,
      light_weapons: 0,
      shield: 0,
      dodge: 0,
      dexterity: 0
    },
    inventory: [],
    equipment: {
      weapon: new Item(items.rusty_large_axe),
      shield: {},
      head: {},
      body: {},
      legs: {},
      ring: {}
    },
    level: {
      lvl: 1,
      pointsPerLvl: 3,
      xpGive: 180
    },
    abilities: [
      new Ability(Abilities.regular_attack),
      new Ability(Abilities.sharp_stroke)
    ],
    statuses: [],
    stat_template: "warrior",
    ai_prefers: {
      regular: 25,
      debuff: 0
    },
    lootTable: "skeleton3"
  },
  "skeleton_mage": {
    id: "skeleton_mage",
    name: "Skeleton Mage",
    sprites: "gfx/enemies/skeleton_mage",
    stats: {
      str: 3,
      vit: 4,
      agi: 8,
      int: 5,
      wis: 5,
      ap: 0,
      hp: 0,
      mp: 0,
      healL: 45,
      critChance: 7,
      critDmg: 55
    },
    resistances: {
      burning: 10,
      bleeding: 0,
      stun: 10,
      power: 0,
      defense: 0,
      freezing: 10
    },
    skills: {
      armorer: 0,
      heavy_weapons: 0,
      light_weapons: 5,
      shield: 0,
      dodge: 5,
      dexterity: 0
    },
    inventory: [],
    equipment: {
      weapon: new Item(items.magician_orb_staff),
      shield: {},
      head: {},
      body: {},
      legs: {},
      ring: {}
    },
    level: {
      lvl: 1,
      pointsPerLvl: 3,
      xpGive: 200
    },
    abilities: [
      new Ability(Abilities.mana_blast),
      new Ability(Abilities.flame_hand)
    ],
    statuses: [],
    stat_template: "mage",
    ai_prefers: {
      regular: 0,
      debuff: 25
    },
    lootTable: "skeleton3"
  }
}

const lootTables = {
  skeleton1: [
    {item: "wooden_shield", chance: 37},
    {item: "rusty_short_sword", chance: 41},
    {item: "weak_healing_gem", chance: 26, amount: [1, 2]},
    {item: "gold", chance: 100, amount: [7, 13]}
  ],
  skeleton2: [
    {item: "bent_wooden_bow", chance: 38},
    {item: "weak_healing_gem", chance: 28, amount: [1, 2]},
    {item: "gold", chance: 100, amount: [11, 22]}
  ],
  skeleton3: [
    {item: "rusty_large_axe", chance: 43},
    {item: "ring_of_woes", chance: 12},
    {item: "weak_healing_gem", chance: 28, amount: [1, 3]},
    {item: "gold", chance: 100, amount: [23, 37]}
  ]
}